File: TRPO/lab5/scala_lab5/src/main/scala/structures/SinglyLinkedList.scala
//package structures

class SinglyLinkedList(userType: UserType) {
  private case class Node(var data: Any, var next: Node = null)
  private var head: Node = null
  private var size = 0

  // Добавление в конец
  def add(value: Any): Unit = {
    val newNode = Node(userType.clone(value))
    if (head == null) head = newNode
    else {
      var current = head
      while (current.next != null) current = current.next
      current.next = newNode
    }
    size += 1
  }

  // Сортировка слиянием (функциональный стиль)
  def functionalSort: SinglyLinkedList = {
    def mergeSort(list: List[Any]): List[Any] = list match {
      case Nil => list
      case _ :: Nil => list
      case _ =>
        val (left, right) = list.splitAt(list.length / 2)
        merge(mergeSort(left), mergeSort(right))
    }

    def merge(left: List[Any], right: List[Any]): List[Any] = (left, right) match {
      case (Nil, _) => right
      case (_, Nil) => left
      case (l :: ls, r :: rs) =>
        if (userType.getTypeComparator.compare(l, r) < 0) l :: merge(ls, right)
        else r :: merge(left, rs)
    }

    val sortedList = new SinglyLinkedList(userType)
    mergeSort(toList).foreach(sortedList.add)
    sortedList
  }

  // Преобразование в список
  def toList: List[Any] = {
    var result = List.empty[Any]
    var current = head
    while (current != null) {
      result = result :+ current.data
      current = current.next
    }
    result
  }
}

File: TRPO/lab6/project-root/bin/main/SinglyLinkedList.kt
import java.io.Serializable

class SinglyLinkedList(val userType: UserType) : Serializable {
    private data class Node(var data: Any, var next: Node? = null)
    private var head: Node? = null
    private var size = 0

    // Добавление элемента
    fun add(value: Any) {
        val cloned = userType.clone(value)
        val newNode = Node(cloned)
        if (head == null) {
            head = newNode
        } else {
            var current = head
            while (current?.next != null) {
                current = current.next
            }
            current?.next = newNode
        }
        size++
    }

    // Получение элемента по индексу
    fun get(index: Int): Any? {
        require(index in 0 until size) { "Индекс $index вне диапазона" }
        var current = head
        repeat(index) { current = current?.next }
        return current?.data
    }

    // Удаление элемента
    fun remove(index: Int): Boolean {
        if (index !in 0 until size) return false
        if (index == 0) {
            head = head?.next
        } else {
            var prev = head
            repeat(index - 1) { prev = prev?.next }
            prev?.next = prev?.next?.next
        }
        size--
        return true
    }

    // Сортировка (слиянием)
    fun sort() {
        head = mergeSort(head)
    }

    private fun mergeSort(head: Node?): Node? {
        if (head?.next == null) return head
        val middle = getMiddle(head)
        val nextToMiddle = middle?.next
        middle?.next = null
        return merge(mergeSort(head), mergeSort(nextToMiddle))
    }

    private fun merge(left: Node?, right: Node?): Node? {
        val comparator = userType.getComparator()
        val dummy = Node(0)
        var tail = dummy
        var l = left
        var r = right

        while (l != null && r != null) {
            if (comparator.compare(l.data, r.data) < 0) {
                tail.next = l
                l = l.next
            } else {
                tail.next = r
                r = r.next
            }
            tail = tail.next!!
        }
        tail.next = l ?: r
        return dummy.next
    }

    private fun getMiddle(head: Node?): Node? {
        var slow = head
        var fast = head
        while (fast?.next?.next != null) {
            slow = slow?.next
            fast = fast.next?.next
        }
        return slow
    }

    fun toList(): List<Any> {
        val result = mutableListOf<Any>()
        var current = head
        while (current != null) {
            result.add(current.data)
            current = current.next
        }
        return result
    }

    fun saveToFile(filename: String) {
        File(filename).writeText(this.toList().joinToString("\n") { 
            userType.parseValue(it.toString()).toString() 
        })
    }

    fun loadFromFile(filename: String) {
        File(filename).readLines().forEach { 
            add(userType.parseValue(it)) 
        }
    }
}